api/util/collections.hpp

```diff
@@ -0,0 +1,644 @@
+  return result;
+}
+
+///
+/// Operator to sort the contents of the specified sequence
+///
+/// @param data_set
+///   The sequence to sort
+///
+/// @return A new sorted sequence
+///
+template<typename T,
+         typename = std::enable_if_t<
+                    has_random_iterator<
+                    std::remove_cv_t<std::remove_reference_t<T>>>::value>>
+auto operator|(T&& data_set, const struct sort_tag&) {
+  auto result = std::forward<T>(data_set);
+  std::sort(std::begin(result), std::end(result));
+  return result;
+}
+
+///
+/// Operator to sort the contents of the specified sequence
+///
+/// This version preserves the order of equal elements
+///
+/// @param data_set
+///   The sequence to sort
+///
+/// @return A new sorted sequence
+///
+template<typename T,
+         typename = std::enable_if_t<
+                    has_random_iterator<
+                    std::remove_cv_t<std::remove_reference_t<T>>>::value>>
+auto operator|(T&& data_set, const struct stable_sort_tag&) {
+  auto result = std::forward<T>(data_set);
+  std::stable_sort(std::begin(result), std::end(result));
+  return result;
+}
+
+///
+/// Operator to remove duplicated elements in the specified sequence
+///
+/// @param data_set
+///   The sequence to remove duplicated elements from
+///
+/// @return A new sequence with unique elements
+///
+template<typename T,
+         typename = std::enable_if_t<
+                    has_random_iterator<
+                    std::remove_cv_t<std::remove_reference_t<T>>>::value>>
+auto operator|(T&& data_set, const struct unique_tag&) {
+  auto result  = (std::forward<T>(data_set) | stable_sort);
+  const auto _ = std::unique(std::begin(result), std::end(result));
+  result.erase(_, result.cend());
+  return result;
+}
+
+///
+/// Operator to sum the contents in the specified sequence
+///
+/// @param data_set
+///   The sequence to sum the contents of
+///
+/// @return The sum of the contents in the specified sequence
+///
+/// @note The elements in the specified sequence must support
+/// the arithmetic operator `+`
+///
+template<typename T,
+         typename = std::enable_if_t<
+                    is_sequence<
+                    std::remove_cv_t<std::remove_reference_t<T>>>::value>>
+auto operator|(T&& data_set, const struct sum_tag&) {
+  using sequence = std::remove_reference_t<T>;
+  return std::accumulate(std::begin(data_set), std::end(data_set), typename sequence::value_type{});
+}
+
+///
+/// Extract the keys from a std::map
+///
+/// @param map
+///   The std::map to extract the keys from
+///
+/// @param reverse_order
+///   If true keys are retrieved in reverse order
+///
+/// @return A std::vector containing all the keys from the
+/// specified std::map
+///
+template<typename K, typename V, typename C, typename A>
+auto keys(const std::map<K,V,C,A>& map, const bool reverse_order = false) {
+  std::vector<K> keys;
+  keys.reserve(map.size());
+  for (const auto& _ : map) keys.push_back(_.first);
```

Comment:
How about using classic for-loop with `std::map::rbegin` in case of `reverse_order == true` to avoid calling `std::reverse()` below?